Sources/ACKLocalizationCore/ACKLocalization.swift

```diff
@@ -0,0 +1,236 @@
+//
+//  ACKLocalization.swift
+//  
+//
+//  Created by Jakub Olejník on 11/12/2019.
+//
+
+import Combine
+import Foundation
+
+public typealias MappedValues = [String: [LocRow]]
+
+public final class ACKLocalization {
+    private let authAPI: AuthAPIServicing
+    private let sheetsAPI: SheetsAPIServicing
+    private var fetchCancellable: Cancellable?
+    
+    // MARK: - Initializers
+    
+    public init(authAPI: AuthAPIServicing = AuthAPIService(), sheetsAPI: SheetsAPIServicing = SheetsAPIService()) {
+        self.authAPI = authAPI
+        self.sheetsAPI = sheetsAPI
+    }
+    
+    // MARK: - Public interface
+    
+    public func run() {
+        let dispatchGroup = DispatchGroup()
+        
+        dispatchGroup.enter()
+        do {
+            let config = try loadConfiguration()
+            
+            fetchCancellable = fetchSheetValues(config)
+                .flatMap { [weak self] in self?.transformValuesPublisher($0, with: config) ?? Fail(error: LocalizationError(message: "Unable to transform values")).eraseToAnyPublisher() }
+                .flatMap { [weak self] in self?.saveMappedValuesPublisher($0, config: config) ?? Fail(error: LocalizationError(message: "Unable to save mapped values")).eraseToAnyPublisher() }
+                .sink(receiveCompletion: { [weak self] result in
+                    switch result {
+                    case .failure(let error):
+                        self?.displayError(error)
+                        exit(1)
+                    case .finished: self?.displaySuccess()
+                    }
+                    dispatchGroup.leave()
+                }) { _ in }
+        } catch {
+          switch error {
+            case let localizationError as LocalizationError:
+                displayError(localizationError)
+            default:
+                print(error)
+            }
+            exit(1)
+        }
+        
+        dispatchGroup.wait()
+    }
+    
+    public func fetchSheetValues(_ spreadsheetTabName: String?, spreadsheetId: String, serviceAccount: ServiceAccount) -> AnyPublisher<ValueRange, LocalizationError> {
+        let sheetsAPI = self.sheetsAPI
+        var accessToken: AccessToken?
+        
+        return authAPI.fetchAccessToken(serviceAccount: serviceAccount)
+            .handleEvents(receiveOutput: { accessToken = $0 })
+            .flatMap { sheetsAPI.fetchSpreadsheet(spreadsheetId, accessToken: $0) }
+            .flatMap { sheetsAPI.fetchSheet(spreadsheetTabName, from: $0, accessToken: accessToken!) }
```

Comment:
Well it might use a comment, but if it is not set by the `handleEvents` side-effect then it means that `fetchAccessToken` failed, hence this code would be skipped, wouldn't it?

Reply:
You are right it will never be nil, but I think it might not be obvious why it is handled the way it is to future devs - so yeah, at least provide a comment, but I still think passing the value would be a better solution since it's more safe and readable anyway.

Reply:
Well I think it is now rather elegantly solved by moving the `accessToken` variable to `SheetsAPIService` 🙂

Reply:
I like that much better, great 🎉